src/protocols/gwmp/hpr_gwmp_worker.erl

```diff
@@ -0,0 +1,500 @@
+    {noreply, State}.
+
+%% @private
+%% @doc This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any
+%% necessary cleaning up. When it returns, the gen_server terminates
+%% with Reason. The return value is ignored.
+-spec terminate(
+    Reason :: (normal | shutdown | {shutdown, term()} | term()),
+    State :: #state{}
+) -> term().
+terminate(_Reason, _State = #state{socket = Socket}) ->
+    ok = gwmp_udp_socket:close(Socket).
+
+%% @private
+%% @doc Convert process state when code is changed
+-spec code_change(
+    OldVsn :: term() | {down, term()},
+    State :: #state{},
+    Extra :: term()
+) ->
+    {ok, NewState :: #state{}} | {error, Reason :: term()}.
+code_change(_OldVsn, State = #state{}, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+hpr_gwmp_send_response_function(Handler) ->
+    fun(Data) ->
+        PacketDown = hpr_gwmp_router:txpk_to_packet_down(Data),
+        grpcbox_stream:send(false, PacketDown, Handler),
+        lager:info("hpr_gwmp_send_response: Data: ~p, Handler: ~p", [Data, Handler])
+    end.
+
+-spec handle_udp(
+    Data :: binary(),
+    DataSrc :: gwmp_udp_socket:socket_dest(),
+    State :: #state{}
+) -> {noreply, #state{}}.
+handle_udp(
+    Data,
+    DataSrc,
+    #state{
+        push_data = PushData0,
+        pull_data_timer = PullDataTimer,
+        pull_data = PullDataMap0,
+        socket = Socket,
+        pull_resp_fun = PullRespFunction,
+        pubkeybin = PubKeyBin
+    } = State0
+) ->
+    State1 =
+        case semtech_udp:identifier(Data) of
+            ?PUSH_ACK ->
+                PushData1 = handle_push_ack(Data, PushData0),
+                State0#state{push_data = PushData1};
+            ?PULL_ACK ->
+                PullDataMap1 = handle_pull_ack(Data, DataSrc, PullDataMap0, PullDataTimer),
+                State0#state{pull_data = PullDataMap1};
+            ?PULL_RESP ->
+                ok = handle_pull_resp(Data, PubKeyBin, Socket, PullRespFunction),
+                State0;
+            _Id ->
+                lager:warning("got unknown identifier ~p for ~p", [_Id, Data]),
+                State0
+        end,
+    {noreply, State1}.
+
+-spec pubkeybin_to_mac(binary()) -> binary().
+pubkeybin_to_mac(PubKeyBin) ->
+    <<(xxhash:hash64(PubKeyBin)):64/unsigned-integer>>.
+
+-spec schedule_pull_data(non_neg_integer(), gwmp_udp_socket:socket_dest()) -> reference().
+schedule_pull_data(PullDataTimer, SocketDest) ->
+    _ = erlang:send_after(PullDataTimer, self(), {?PULL_DATA_TICK, SocketDest}).
+
+-spec schedule_shutdown(non_neg_integer()) -> reference().
+schedule_shutdown(ShutdownTimer) ->
+    _ = erlang:send_after(ShutdownTimer, self(), ?SHUTDOWN_TICK).
+
+update_address(Socket0, Address, Port) ->
```

Comment:
I'm wondering if we still need the `gwmp_udp_socket`. 
The only feature we're really using from it is keeping the location with the socket, but we're back in a position where we have to update the address before sending anything. 

It might be nice to keep the module for the types it defines, but removing `update_address/2` and requiring the dest be supplied with each `send/3`.